File: app/src/main/java/mb00/android/codehub/api/parser/FileSizeParser.java
package mb00.android.codehub.api.parser;

import java.math.BigDecimal;
import java.math.MathContext;


public class FileSizeParser {

    public static String parseSize(String size) {
        if (size.length() <= 3) { // size in bytes
            size += "B";
            return size;
        } else if (size.length() == 4) { // size in kilobytes
            String kilobytesInteger = size.substring(0, 1);
            String kilobytesFractional = size.substring(1);
            String sizeInKilobytes = kilobytesInteger + "." + kilobytesFractional;
            size = new BigDecimal(sizeInKilobytes).round(new MathContext(2)).toString() + "KB";
            return size;
        } else if (size.length() == 5) { // size in kilobytes
            String kilobytesInteger = size.substring(0, 2);
            String kilobytesFractional = size.substring(2);
            String sizeInKilobytes = kilobytesInteger + "." + kilobytesFractional;
            size = new BigDecimal(sizeInKilobytes).round(new MathContext(2)).toString() + "KB";
            return size;
        } else if (size.length() == 6){ // size in megabytes;
            String megabytesInteger = size.substring(0, 1);
            String megabytesFractional = size.substring(1);
            String sizeInMegabytes = megabytesInteger + "." + megabytesFractional;
            size = new BigDecimal(sizeInMegabytes).round(new MathContext(2)).toString() + "MB";
            return size;
        } else { // some horrific edge case
            return size;
        }
    }

}

File: app/src/main/java/mb00/android/codehub/logic/utils/FileSizeParser.kt
package mb00.android.codehub.logic.utils

import java.math.BigDecimal
import java.math.MathContext

/**
 * Contains methods used to parse file sizes from the GitHub API
 */

object FileSizeParser {

    fun parseSize(size: String?): String? {
        size?.let {
            return when {
                size.length <= 3 -> returnBytes(size)
                size.length == 4 || size.length == 5 -> returnKilobytes(size)
                size.length == 6 -> returnMegabytes(size)
                else -> size // some horrific edge case
            }
        }
    }

    private fun returnBytes(size: String?): String? {
        return "$size B"
    }

    private fun returnKilobytes(size: String?): String? {
        val kilobytesInteger = size?.substring(0, size.length - 3)
        val kilobytesFractional = size?.substring(size.length - 3)
        val sizeInKilobytes = "$kilobytesInteger.$kilobytesFractional"
        return BigDecimal(sizeInKilobytes).round(MathContext(2)).toString() + " KB"
    }

    private fun returnMegabytes(size: String?): String? {
        val megabytesInteger = size?.substring(0, 1)
        val megabytesFractional = size?.substring(1)
        val sizeInMegabytes = "$megabytesInteger.$megabytesFractional"
        return BigDecimal(sizeInMegabytes).round(MathContext(2)).toString() + " MB"
    }

}
